### PR TITLE
gcc: patch for https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100102

### PR DESCRIPTION
**Pull request by danielahlin**
Fixes a bug which causes gcc to crash when used by nvcc to build
e.g. py-torch with CUDA/cudnn. The patch is the one provided in
the bug URL.

### CODE REVIEW

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 1]
```diff
@@ -287,6 +287,9 @@ class Gcc(AutotoolsPackage, GNUMirrorPackage):
     # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=95005
     patch('zstd.patch', when='@10')
 
+    # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100102
+    patch('ICE_in_tsubst.patch', when='@10.3.0')
+
```
Comment (michaelkuhn) [inline var/spack/repos/builtin/packages/gcc/package.py]: Three comments:
1. Could we add those via URL patches so we do not have to carry the patch ourselves? Should be https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=fc930b3010bd0de899a3da3209eab20664ddb703 for GCC 10 and https://gcc.gnu.org/git/?p=gcc.git;a=patch;h=f1feb74046e0feb0596b93bbb822fae02940a90e for GCC 11.
2. Why is this only applied for 10.3.0? 10.1 and 10.2 are probably also affected, right?
3. We should also patch 11.1, which is also affected according to the upstream bug.

**Review by michaelkuhn — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gcc/package.py (hunk 1 above)
Comment (danielahlin) [reply var/spack/repos/builtin/packages/gcc/package.py]: Thanks for the feedback.

1. Ok - sounds good.
2&3. I had only tested with 10.3.0 - but you are right the other ones should be updated as well.

I've updated the PR as suggested in a new commit.

**Comment by michaelkuhn** [comment]
Sorry for dropping the ball on this one. Could you please resolve the conflict? I'll merge it then. Thanks!

**Comment by danielahlin** [comment]
Conflict was due to https://github.com/spack/spack/pull/25980 which fixes the same issue.

I've resolved conflict by preferring this pull request as:

1. https://github.com/spack/spack/pull/25980 only applies to 10.3 and 11.1 while the bug is likely present in 10.1:10.3
2. https://github.com/spack/spack/pull/25980 uses the same patch for 10 and 11 while https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100102 are using different patches for 10 and 11 (I have not checked if it really makes a difference or not though).

**Review by michaelkuhn — APPROVED** [review]

**Pull request merged**